src/backend/vaapi.rs

```diff
@@ -224,6 +210,56 @@ mod surface_pool {
 
     use crate::Resolution;
 
+    /// A VA Surface obtained from a `[SurfacePool]`.
+    ///
+    /// The surface will automatically be returned to its pool upon dropping, provided the pool still
+    /// exists and the surface is still compatible with it.
+    pub struct PooledSurface<D: SurfaceMemoryDescriptor> {
```

Comment:
This looks like a reimplementation of what `GenericBackendHandler` already provided?

I wonder if ownership wasn't more clear that way, i.e.:

1. Surface is initially owned by the pool,
2. Ownership is transferred when the surface is acquired through `get_surface()`
3. Optionally, a type can return the surface to the pool when going out of scope

IMHO (so far, I haven't seen the rest of the series yet), this achieves the same in a more complex way. Also, you can't "leak" the resource, it will simply drop.

Reply:
There are several reasons for this, the main one is that for future work on surface allocations (esp. regarding external memory) I need the pool to ultimately own the surfaces and be able to account for how many it is currently managing. The former `get_surface` implementation did not allow that as the surface ownership was fully transferred to the owner, and we relied on its goodwill to return it later.

The `PooledSurface` type is a type that does one thing, return the surface to the pool when it is dropped (with a possibility to fully take ownership too, but that's opt-in). In that respect I think it is a bit easier to grasp than the previous implementation, which relies on `TryInto<Surface<D>> for PictureState` and code in other types to work. Here everything is in the same place.

The point of this should hopefully be clearer with the following CLs of this series.